pkg/serviceaccount/ensure.go

```diff
@@ -20,28 +20,154 @@ package serviceaccount
 
 import (
 	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"time"
 
+	"github.com/pkg/errors"
 	"github.com/submariner-io/admiral/pkg/resource"
 	resourceutil "github.com/submariner-io/subctl/pkg/resource"
+	"github.com/submariner-io/subctl/pkg/secret"
 	"github.com/submariner-io/submariner-operator/pkg/embeddedyamls"
 	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/kubernetes"
 )
 
-// EnsureFromYAML creates the given service account.
+const (
+	createdByAnnotation = "kubernetes.io/created-by"
+	creatorName         = "subctl"
+)
+
+// ensureFromYAML creates the given service account.
 // nolint:wrapcheck // No need to wrap errors here.
-func EnsureFromYAML(kubeClient kubernetes.Interface, namespace, yaml string) (bool, error) {
+func ensureFromYAML(kubeClient kubernetes.Interface, namespace, yaml string) (*corev1.ServiceAccount, error) {
 	sa := &corev1.ServiceAccount{}
 
 	err := embeddedyamls.GetObject(yaml, sa)
 	if err != nil {
-		return false, err
+		return nil, err
+	}
+
+	_, err = Ensure(kubeClient, namespace, sa)
+	if err != nil {
+		return nil, err
 	}
 
-	return Ensure(kubeClient, namespace, sa)
+	return sa, err
 }
 
 // nolint:wrapcheck // No need to wrap errors here.
 func Ensure(kubeClient kubernetes.Interface, namespace string, sa *corev1.ServiceAccount) (bool, error) {
 	return resourceutil.CreateOrUpdate(context.TODO(), resource.ForServiceAccount(kubeClient, namespace), sa)
 }
+
+// EnsureFromYAMLWithSecret creates the given service account and secret for it.
+func EnsureFromYAMLWithSecret(kubeClient kubernetes.Interface, namespace, yaml string) (bool, error) {
+	sa, err := ensureFromYAML(kubeClient, namespace, yaml)
+	if err != nil {
+		return false, errors.Wrap(err, "error provisioning the ServiceAccount resource")
+	}
+
+	saSecret, err := EnsureSecretFromSA(kubeClient, sa.Name, namespace)
+	if err != nil {
+		return false, errors.Wrap(err, "error creating secret for ServiceAccount resource")
+	}
+
+	return sa != nil && saSecret != nil, nil
+}
+
+func EnsureSecretFromSA(client kubernetes.Interface, saName, namespace string) (*corev1.Secret, error) {
+	sa, err := client.CoreV1().ServiceAccounts(namespace).Get(context.TODO(), saName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get ServiceAccount %s/%s", namespace, saName)
+	}
+
+	saSecret := getSecretFromSA(client, sa)
+
+	if saSecret != nil {
+		return saSecret, nil
+	}
+
+	// We couldn't find right secret from this SA, search all Secrets
+	saSecret, err = getSecretForSA(client, sa)
+	if err != nil {
```

Comment:
We should distinguish the `NotFound` error.

```suggestion
	if err != nil && !apierrors.IsNotFound(err) {
	    return err
	}
	
	if err != nil {
	
```